CMakeLists.txt

```diff
@@ -174,9 +176,52 @@ foreach(src_file ${uhdm-GENERATED_SRC})
   set_source_files_properties(${src_file} PROPERTIES GENERATED TRUE)
 endforeach(src_file ${uhdm-GENERATED_SRC})
 
+# Parse out version and configure header
+if(EXISTS "${CMAKE_SOURCE_DIR}/.git")
+  # running a build from the git repo, parse out tag and sha
+  # get latest commit
+  execute_process(COMMAND git describe --tags --abbrev=0
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    OUTPUT_VARIABLE LATEST_VERSION)
+
+  # parse out first number after "v"
+  string(REGEX REPLACE "^v([0-9]+)\\..*" "\\1" UHDM_VERSION_MAJOR "${LATEST_VERSION}")
+  # parse out second number, e.g. after "v[0-9]."
+  string(REGEX REPLACE "^v[0-9]+\\.([0-9]+).*" "\\1" UHDM_VERSION_MINOR "${LATEST_VERSION}")
+
+  # get latest commit
+  execute_process(COMMAND git rev-parse HEAD
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    OUTPUT_VARIABLE UHDM_VERSION_COMMIT_SHA)
+  # strip newline
+  string(REGEX REPLACE "\n$" "" UHDM_VERSION_COMMIT_SHA "${UHDM_VERSION_COMMIT_SHA}")
+
+  if("${UHDM_VERSION_MAJOR}" STREQUAL "")
+    # No tags locally, user project version
+    set(UHDM_VERSION_MAJOR ${CMAKE_PROJECT_VERSION_MAJOR})
+    set(UHDM_VERSION_MINOR ${CMAKE_PROJECT_VERSION_MINOR})
+  elseif(NOT "${UHDM_VERSION_MAJOR}.${UHDM_VERSION_MINOR}" STREQUAL "${CMAKE_PROJECT_VERSION_MAJOR}.${CMAKE_PROJECT_VERSION_MINOR}")
+    message(FATAL_ERROR "Git tag ${UHDM_VERSION_MAJOR}.${UHDM_VERSION_MINOR} \
+      does not match project version ${CMAKE_PROJECT_VERSION_MAJOR}.${CMAKE_PROJECT_VERSION_MINOR}. \
+      Did you forget to update project version in CMakeLists.txt?")
+  endif()
+else()
+  set(UHDM_VERSION_MAJOR ${CMAKE_PROJECT_VERSION_MAJOR})
+  set(UHDM_VERSION_MINOR ${CMAKE_PROJECT_VERSION_MINOR})
+  set(UHDM_VERSION_COMMIT_SHA "release")
+endif()
+
+set(UHDM_BUILD_TYPE ${CMAKE_BUILD_TYPE})
+message("Building version v${UHDM_VERSION_MAJOR}.${UHDM_VERSION_MINOR} [${UHDM_VERSION_COMMIT_SHA}]")
+configure_file(${CMAKE_SOURCE_DIR}/include/uhdm-version.h.in ${GENDIR}/uhdm/uhdm-version.h)
+
 add_library(uhdm ${uhdm-GENERATED_SRC})
 set_target_properties(uhdm PROPERTIES PUBLIC_HEADER ${GENDIR}/uhdm/uhdm.h)
 
+if(BUILD_SHARED_LIBS)
```

Comment:
Does this allow to properly compile uhdm as shared library with in particular `capnproto ` to be compile with `-fPIC` switch ?

At my end I try to work on a uhdm python wrapper https://github.com/Thomasb81/UHDM/tree/python_wrapper
But if I mimic correctly this in my working environment it failed to properly compile (I may be wrong at doing this):
```
[100%] Built target uhdm-hier
[100%] Linking CXX shared module ../bin/_swig_example.so
/usr/bin/ld: ../third_party/capnproto/c++/src/kj/libkj.a(exception.c++.o): relocation R_X86_64_TPOFF32 against `_ZN2kj12_GLOBAL__N_1L19threadLocalCallbackE' can not be used when making a shared object; recompilé avec -fPIC
```

I see in antlr project that instead to control static/shared library they define 2 targets one for static and the second for shared. This allow end application to choose.
The drawback is project build without specific target will compile twice the library.

Is there arguments to continue to link uhdm against static library ? It seems to me that linking against shared library could be more efficient regarding elaboration time.

Reply:
Yes, this PR lets you compile UHDM as a shared library via the cmake flag `BUILD_SHARED_LIBS`. Capnproto and the other dependencies have their own separate build processes. However, you can see how now when I build `capnproto` in the `main.yml`, I use the cmake flag `CMAKE_POSITION_INDEPENDENT_CODE` which will use the compiler-specific flag (e.g. `-fPIC`). If you want to do a build with vendored sources (e.g. from the `git` repo) you would run a command like: `cmake -B build -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_POSITION_INDEPENDENT_CODE=ON . && cmake --build build`.